### PR TITLE
[WebGPU] Async, loop-based invoke and output.

### PR DESCRIPTION
**Pull request by ScottTodd**
Now that https://github.com/openxla/iree/pull/13669 added support for `iree_loop_wait_all()` to `iree_loop_emscripten`, this builds on the work that https://github.com/openxla/iree/pull/11017 started to convert the sample WebGPU application to be fully asynchronous[1].

The code _should_ support multiple return values, but I ran into https://github.com/openxla/iree/issues/13809 while working on this.

^[1]: `iree_hal_semaphore_wait` is still technically synchronous, see the notes in the source code

---

Draft, TODO:

- [ ] populate `async_status`
- [ ] heap allocate `wait_sources` array if needed
- [ ] plumb output statuses up to JS? Emscripten asserts just crash the page, so our status tracking can't report everything...

### CODE REVIEW

File: experimental/web/sample_webgpu/main.c [hunk 1]
```diff
@@ -546,16 +551,30 @@ static iree_status_t print_outputs_from_call(
...
+          &heap_buffer_view));
+      // Query total length (excluding NUL terminator).
+      iree_host_size_t result_length = 0;
+      iree_status_t status = iree_hal_buffer_view_format(
+          heap_buffer_view, SIZE_MAX, 0, NULL, &result_length);
+      if (!iree_status_is_out_of_range(status)) return status;
+      status = iree_ok_status();
+      ++result_length;  // include NUL
+      // Format into a scratch buffer and append.
+      char* scratch = (char*)iree_alloca(result_length);
```
Comment (benvanik) [inline experimental/web/sample_webgpu/main.c]: this is dangerous - alloca only really works for a small number of KB of allocations (dozens of bytes to 1-4KB)

instead, you can use iree_string_builder_append_inline - you compute the required length like you do, call that with the length+1 (NUL) to get the char* pointer, then iree_hal_buffer_view_format right into the storage.

File: experimental/web/sample_webgpu/main.c [hunk 2]
```diff
@@ -572,28 +591,302 @@ static iree_status_t print_outputs_from_call(
...
+    // Note: this leaks the buffer. It's up to the caller to free it after use.
+    char* outputs_string =
+        strdup(iree_string_builder_buffer(&output_string_builder));
+    iree_string_builder_deinitialize(&output_string_builder);
+    call_state->callback_fn(outputs_string);
+  } else {
+    fprintf(stderr, "map_all_callback error:\n");
+    iree_status_fprint(stderr, status);
+    // Note: loop_emscripten.js must free 'status'!
+    call_state->callback_fn(0);
```
Comment (benvanik) [inline experimental/web/sample_webgpu/main.c]:
```suggestion
    call_state->callback_fn(NULL);
```

**Review by benvanik — CHANGES_REQUESTED** [review]
1 fix needed then lgtm

File: experimental/web/sample_webgpu/main.c (hunk 1 above)
Comment (ScottTodd) [reply experimental/web/sample_webgpu/main.c]: Even better, I can just call `iree_hal_buffer_view_append_to_builder` that already does that :)

**Comment by ScottTodd** [comment]
PR closed since the base branch was deleted. I'll rebase and open a new PR.

**Pull request closed without merging**